evm/src/arithmetic/byte.rs

```diff
@@ -0,0 +1,502 @@
+    //
+    // but we can't borrow both `tree` and `val_limbs` simultaneously.
+    // Apparently the solution is to use `split_at_mut()`; below we
+    // assume that the `val` registers are earlier in the row than the
+    // `tree` registers, so we enforce that assumption here.
+    let val_idx = INPUT_REGISTER_1.start;
+    let tree_idx = AUX_INPUT_REGISTER_1.start;
+    assert!(val_idx + N_LIMBS < tree_idx);
+
+    // `lvl_len` is the number of elements of the current level of the
+    // "tree". Can think of `val_limbs` as level 0, with length =
+    // N_LIMBS = 16.
+    const_assert!(N_LIMBS == 16); // Enforce assumption
+
+    let (prev, next) = lv[val_idx..].split_at_mut(tree_idx - val_idx);
+    let lvl_len = 8;
+    let offset = (!idx.bit(4) as usize) * lvl_len;
+    next[..lvl_len].copy_from_slice(&prev[offset..offset + lvl_len]);
+
+    let (prev, next) = next.split_at_mut(lvl_len);
+    let lvl_len = 4;
+    let offset = (!idx.bit(3) as usize) * lvl_len;
+    next[..lvl_len].copy_from_slice(&prev[offset..offset + lvl_len]);
+
+    let (prev, next) = next.split_at_mut(lvl_len);
+    let lvl_len = 2;
+    let offset = (!idx.bit(2) as usize) * lvl_len;
+    next[..lvl_len].copy_from_slice(&prev[offset..offset + lvl_len]);
+
+    let (prev, next) = next.split_at_mut(lvl_len);
+    // lvl_len = 1
+    let offset = !idx.bit(1) as usize;
+    next[0] = prev[offset]; // tree[14] = tree[offset + 12]
+
+    // Handle the last bit; i.e. pick a byte of the final limb.
+    let t = next[0].to_canonical_u64();
+    let lo = t as u8 as u64;
+    let hi = t >> 8;
+
+    // Store 256 * lo rather than lo:
+    lv[BYTE_LAST_LIMB_LO] = F::from_canonical_u64(lo << 8);
+    lv[BYTE_LAST_LIMB_HI] = F::from_canonical_u64(hi);
+
+    let tree = &mut lv[AUX_INPUT_REGISTER_1];
+    let output = if idx.bit(0) {
+        tree[15] = F::from_canonical_u64(lo);
+        lo.into()
+    } else {
+        tree[15] = F::from_canonical_u64(hi);
+        hi.into()
+    };
+
+    u256_to_array(
+        &mut lv[OUTPUT_REGISTER],
+        if idx < 32.into() {
+            output
+        } else {
+            U256::zero()
+        },
+    );
+}
+
+pub fn eval_packed<P: PackedField>(
+    lv: &[P; NUM_ARITH_COLUMNS],
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    let is_byte = lv[IS_BYTE];
+
+    let idx = &lv[INPUT_REGISTER_0];
+    let val = &lv[INPUT_REGISTER_1];
+    let out = &lv[OUTPUT_REGISTER];
+    let idx_decomp = &lv[AUX_INPUT_REGISTER_0];
+    let tree = &lv[AUX_INPUT_REGISTER_1];
+
+    // low 5 bits of the first limb of idx:
+    let mut idx0_lo5 = P::ZEROS;
+    for i in 0..5 {
+        let bit = idx_decomp[i];
+        yield_constr.constraint(is_byte * (bit * bit - bit));
+        idx0_lo5 += bit * P::Scalar::from_canonical_u64(1 << i);
+    }
+    // high (11) bits of the first limb of idx:
+    let idx0_hi = idx_decomp[5] * P::Scalar::from_canonical_u64(32u64);
+    yield_constr.constraint(is_byte * (idx[0] - (idx0_lo5 + idx0_hi)));
```

Comment:
I would briefly explain that this guarantees that `idx0_hi` is range-checked to 11-bits.